src/main/scala/wvlet/airframe/SessionImpl.scala

```diff
@@ -57,9 +57,10 @@ private[airframe] class SessionImpl(sessionName:Option[String], binding: Seq[Bin
     val t = ObjectType.ofTypeTag(ev)
     binding.find(_.from == t) match {
       case Some(SingletonBinding(from, to, eager)) =>
-        singletonHolder.getOrElseUpdate(to, {
-          registerInjectee(to, obj)
-        }).asInstanceOf[A]
+        singletonHolder.getOrElseUpdate(from, registerInjectee(to, obj)).asInstanceOf[A]
+      case Some(InstanceBinding(form, obj)) =>
+        // Instance is already registered in init
+        obj.asInstanceOf[A]
```

Comment:
Missing this line was the cause why @PostConstruct is called multiple times for singletons